Add basic test suite setup

jonkoops opened the pull request:
Adds a basic setup for the integration test suite of Keycloak JS. This PR is purposefully minimal to reduce review load, and will be followed up by a set of PRs to add functionality such as CI integration, additional tests, and cross-origin testing.

Closes #53

jonkoops commented on test/app/app.js:
```diff
@@ -0,0 +1,12 @@
+// Set up Express
+const app = express()
+
+// Expose 'public' directory and Keycloak JS source.
+app.use(express.static(path.resolve(import.meta.dirname, 'public')))
+app.use(express.static(path.resolve(import.meta.dirname, '../../lib')))
+
```
I don't care, as this only runs in Node.js at all times.

jonkoops commented on test/app/app.js:
```diff
@@ -0,0 +1,12 @@
+const app = express()
+
+// Expose 'public' directory and Keycloak JS source.
+app.use(express.static(path.resolve(import.meta.dirname, 'public')))
+app.use(express.static(path.resolve(import.meta.dirname, '../../lib')))
+
```
I don't care, as this only runs in Node.js at all times.

jonkoops commented on test/app/app.js:
```diff
@@ -0,0 +1,12 @@
+const app = express()
+
+// Expose 'public' directory and Keycloak JS source.
+app.use(express.static(path.resolve(import.meta.dirname, 'public')))
+app.use(express.static(path.resolve(import.meta.dirname, '../../lib')))
+
```
I don't care, as this only runs in Node.js at all times.

mposolda commented:
@jonkoops Nice!

Few comments:

1) README does not specify expected versions of `npm` and `node` . Should this be specified somehow perhaps? Just FYI. My versions are:

```
mposolda@mposolda-ThinkPad-P1-Gen-3:~/IdeaProjects/keycloak-js$ npm --version
10.9.2
mposolda@mposolda-ThinkPad-P1-Gen-3:~/IdeaProjects/keycloak-js$ node --version
v22.14.0
```

2) When running the tests and the step `npx playwright install`, browsers were downloaded. At the end there was just single warning. Not sure if it could be ignored or if it could mean problem for some browsers? I've just ignored it for now. FYI. my system is Ubuntu 22.04. The warning was:

```
╔══════════════════════════════════════════════════════╗
║ Host system is missing dependencies to run browsers. ║
║ Missing libraries:                                   ║
║     libgstcodecparsers-1.0.so.0                      ║
║     libavif.so.13                                    ║
╚══════════════════════════════════════════════════════╝
    at validateDependenciesLinux (/home/mposolda/IdeaProjects/keycloak-js/node_modules/playwright-core/lib/server/registry/dependencies.js:216:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async Registry._validateHostRequirements (/home/mposolda/IdeaProjects/keycloak-js/node_modules/playwright-core/lib/server/registry/index.js:859:52)
    at async Registry._validateHostRequirementsForExecutableIfNeeded (/home/mposolda/IdeaProjects/keycloak-js/node_modules/playwright-core/lib/server/registry/index.js:957:7)
    at async Registry.validateHostRequirementsForExecutablesIfNeeded (/home/mposolda/IdeaProjects/keycloak-js/node_modules/playwright-core/lib/server/registry/index.js:946:43)
    at async t.<anonymous> (/home/mposolda/IdeaProjects/keycloak-js/node_modules/playwright-core/lib/cli/program.js:122:7)
```


3) When running the step `npm test`, it failed for me. The Keycloak server was started, but for me, it used very old version of Keycloak server (version 21.0.0). As that is probably the one, which I have cached in my `podman` for the image `latest` . I wonder if we want somehow to enforce the proper server version based on the last Keycloak server released version? Or maybe we can add to the README that people are recommended to delete their `latest` image from podman registry before running the test to make sure that latest is downloaded?


4) I've tried to run against my local Keycloak server, which I used during the build. It worked just fine when I've deleted the "gracefulShutdown" step as you specified in the README and when I've did the run of:
```
export KC_HEALTH_ENABLED=true
./kc.sh start-dev
```
This helped to have the test working.

Maybe this could be added to the README as well that people can use their "local" server (with the instructions how to make it working) similarly like we have for admin-console tests in the keycloak repository?

5) Instructions to have single test running can be useful as well if it is possible.

pskopek commented:
> 2. When running the tests and the step `npx playwright install`, browsers were downloaded. At the end there was just single warning. Not sure if it could be ignored or if it could mean problem for some browsers? I've just ignored it for now. FYI. my system is Ubuntu 22.04. The warning was:
> 
> ```
> ╔══════════════════════════════════════════════════════╗
> ║ Host system is missing dependencies to run browsers. ║
> ║ Missing libraries:                                   ║
> ║     libgstcodecparsers-1.0.so.0                      ║
> ║     libavif.so.13                                    ║
> ╚══════════════════════════════════════════════════════╝
>     at validateDependenciesLinux (/home/mposolda/IdeaProjects/keycloak-js/node_modules/playwright-core/lib/server/registry/dependencies.js:216:9)
>     at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
>     at async Registry._validateHostRequirements (/home/mposolda/IdeaProjects/keycloak-js/node_modules/playwright-core/lib/server/registry/index.js:859:52)
>     at async Registry._validateHostRequirementsForExecutableIfNeeded (/home/mposolda/IdeaProjects/keycloak-js/node_modules/playwright-core/lib/server/registry/index.js:957:7)
>     at async Registry.validateHostRequirementsForExecutablesIfNeeded (/home/mposolda/IdeaProjects/keycloak-js/node_modules/playwright-core/lib/server/registry/index.js:946:43)
>     at async t.<anonymous> (/home/mposolda/IdeaProjects/keycloak-js/node_modules/playwright-core/lib/cli/program.js:122:7)
> ```
> 

@mposolda My PR https://github.com/jonkoops/keycloak-js/pull/3 has some additional packages that need to be installed on Ubuntu 22.04 to make Playwright work seamlessly.

pskopek commented:
It looks good for me. I had issues to run playwright on Fedora, but solved it using `distrobox` (see my PR https://github.com/jonkoops/keycloak-js/pull/3). 
I didn't hit the `podman` issues like @mposolda as my environment started from scratch.

I agree with specifying expected versions of `nodejs` and `npm/npx`.

jonkoops commented:
> README does not specify expected versions of npm and node . Should this be specified somehow perhaps? Just FYI.

I added a `(latest LTS or greater)` comment to the prerequisites section, as that is what will be used on CI. This is only for development, so it doesn't matter that much. As for the NPM version, that is shipped with Node.js, so that will be good enough.

> When running the tests and the step npx playwright install, browsers were downloaded. At the end there was just single warning. Not sure if it could be ignored or if it could mean problem for some browsers? I've just ignored it for now. FYI. my system is Ubuntu 22.04. The warning was:

This can be fixed by passing the `--with-deps` flag, I went ahead and added some more information to the guide.

> When running the step npm test, it failed for me. The Keycloak server was started, but for me, it used very old version of Keycloak server (version 21.0.0). As that is probably the one, which I have cached in my podman for the image latest . I wonder if we want somehow to enforce the proper server version based on the last Keycloak server released version?

Using the latest should be fine, I added the `--pull=newer` flag to the Podman command, that will ensure it always pulls the latest if present.

> Maybe this could be added to the README as well that people can use their "local" server (with the instructions how to make it working) similarly like we have for admin-console tests in the keycloak repository?

I've added an alternative version of the command under the 'Speeding up testing' section.

> Instructions to have single test running can be useful as well if it is possible.

Let's handle this as a separate issue if the time comes, this is also already documented in the Playwright documentation, which is linked in the guide already.

pskopek reviewed (approved):
LGTM!

mposolda reviewed (dismissed):
@jonkoops @pskopek Thanks for the improvements! Approving now.

douglaspalmer reviewed (dismissed):
Managed to get it running. LGTM.

douglaspalmer commented on test/README.md:
````diff
@@ -0,0 +1,100 @@
+npx playwright install --with-deps
+```
+
+### Setup on unsupported Linux distributions
````
Maybe this should read: Setup on Linux distributions unsupported by Playwright

jonkoops replied:
I've added some details that this section can be skipped if you are not on an unsupported Linux distro.
